elastic/kibana
# [Logs UI] Prevent summary request from piling up

## :memo: Summary

This prevents the execution of new log stream summary requests until the previous request has finished.

related to #86439

## :female_detective: Review notes

Currently, the `useLogSummary` hook executes new summary requests when any of the parameters change. If parameter changes arrive quicker than the summary requests complete, they will pile up. While the `useCancellableEffect` prevented the requests from overtaking each other, the old in-flight requests would still put load on the cluster.

Ideally we would re-implement the summary fetching as cancellable async searches. But since we're going to change the orchestration of log stream requests soon anyway (as part of #134412), that would be wasteful. So this is an intermediate improvement to reduce the cluster load.

The changes to the `use_observable.ts` collection of hooks are twofold:

- The previous implementation of `useBehaviorSubject` didn't keep the identity of the second return value (the `next` function) stable, which could lead to undesired re-rendering.
- The new `useReplaySubject` hook is similar to the improved `useBehaviorSubject`, but powered by a n RxJS `ReplaySubject`, which doesn't have an initial value.

## :stethoscope: How to test

- visit the Logs UI in a cluster with log entries
- start streaming
- set the refresh interval to a low value like 1s
- throttle the network traffic such that requests to `/api/log_entries/summary` take longer than the refresh interval
- observe that no two requests to that API are ever in flight in parallel
